components/log-inner/write-sync/index.test.mjs

```diff
@@ -1,24 +1,36 @@
-const { encodeURIComponent } = globalThis;
+const {
+  String,
+  URL,
+  JSON: { stringify: stringifyJSON },
+} = globalThis;
 
+import { writeFile as writeFileAsync } from "node:fs/promises";
 import "../../__fixture__.mjs";
-
 import { getUuid } from "../../uuid/random/index.mjs";
-
 import { getTmpUrl } from "../../path/index.mjs";
-
 import { toAbsoluteUrl } from "../../url/index.mjs";
 
-const testAsync = async (url) => {
-  const { logInfo } = await import(url);
+import { logInfo } from "./index.mjs";
+logInfo("message");
+
+let counter = 0;
+
+const testAsync = async (output) => {
+  await writeFileAsync(
+    new URL(toAbsoluteUrl("../../params.mjs", import.meta.url)),
+    `export default {env:"test", "log-file":${stringifyJSON(output)}};`,
+    "utf8",
+  );
```

Comment:
The problem is that these parameters are not constant. First they have a default value before loading the configuration file. Then they should be overwritten based on the configuration. And I wanted to avoid states in modules. The only other solution I could come up with is to pass them around. But that would be pretty verbose as they are pretty much used everywhere since they are needed for logging. These parameters are the only reason why I implemented this on-demand bundling. Without them, I could pre-bundle the components entry-points. So you are making me second guess my decision to not pass them around. Do you have an alternative solution that does not putting states in modules?

Reply:
You've obviously given it much more thought than I did, but my first instinct would be to use environment variables — they're external (so modules are kept stateless), ephemeral yet persistent for the duration of the execution, automatically passed around by the operating system, can be easily changed and as a bonus they'd be really easy for the user to override (just don't set a variable if it's already set before launch).

Reply:
Vue uses an environment variable for conditional 'compilation': https://github.com/vuejs/vue/blob/898d1e19fc3c0a8e899767a627c2d0a09615db9d/dist/vue.common.js — although I suppose in a much more limited manner and it doesn't actually change it around on its own.

Reply:
I agree that environment variables would be a good fit for values that do not change during the entire lifetime of the process. And most of my parameters don't. But unfortunately, some do. In that case, environment variables become nothing more than a global state. And, having a module rely on a global state is arguably worse than having the state encapsulated by the module. Plus, I don't see how environment variables will help propagate the change to said state. Consider the file that spawned this discussion. On loading the module, it opens a file descriptor that will be written onto by subsequent log calls. Initially that file descriptor should be `stdout` then, once the configuration is loaded, it should be based on `config.log.file`. Maybe I try too hard to avoid state and maybe I should use some dirty event handlers. I don't know, I think I would prefer explicitelly passing around these parameters rather than introducing global state. PS I drew insperation from this: https://wiki.haskell.org/Global_variables

Reply:
Regarding this particular example, you could do something like 
```javascript
// log.js
process.env.APPMAP_LOG_FD = 1;

export const log = (...) => writeSync(process.env.APPMAP_LOG_FD, ...);

// config.js
if (config.log_file) process.env.APPMAP_LOG_FD = openSync(config.log_file);
```

or

```javascript
// context.js
let state = {
  logFd: 1
};

export logFd = () => state.logFd;
export openLogFile = (path) => state.logFd = openSync(path);
```

Anyway, params module is still global state. If you want to avoid having many stateful modules, why not just have one context module that holds all the state? (I suppose that's what the params module is...) Recompiling and rebundling based on parameters is still global state :) I guess I'm not sure what's the problem you're trying to solve. If you want it to be really stateless, you should pass the context around everywhere and do things like
```javascript
// foo.js
import logging from 'logging';

export default (params) => {
  const log = logging(params);
  
  return {
    foo: (...) => { 
      ...
      log.warn('foo has been called');
    }
  };
};
```
but at this point you're just fighting with the language trying to get it to do stuff it's not really designed for (in particular ESM).

Reply:
I did as you suggested and move to static bundling. It think it is much better now. The bulk of the change happened here: bfbc0005ecdd176802173f1c852027faa1043e55.